content/blog/2019-09-25-bril-c-backend.md

```diff
@@ -0,0 +1,92 @@
+
+In order to verify the correctness of this C backend, I create some valid handwritten tests to verify against the existing interpreter `brili`. The C backend successfully passes all the tests.
+
+Here we show a small example to demonstrate the translation.
+```
+main {
+  a: int = const 4;
+  b: int = const 4;
+  cmp: bool = ge a b;
+  jmp somewhere;
+  a: int = const 2;
+  somewhere:
+  c: int = add a b;
+  print c;
+  print cmp;
+}
+```
+The above Bril program can be translate to C as the following.
+```C
+#include <stdint.h>
+#include <stdio.h>
+#include <inttypes.h>
+int main(){
+int64_t a;
+int64_t b;
+int cmp;
+int64_t c;
+a = 4LL;
+b = 4LL;
+cmp = a >= b;
+goto somewhere;
+a = 2LL;
+somewhere:;
+c = a + b;
+printf("%" PRId64 "\n", c);
+printf(cmp?"true\n":"false\n");
+return 0;
+}
+```
+
+Implementation
+---
+I implemented the translation tool in Python. The implementation is straightforward and it consists of 134 lines of code.  Compared to other projects such as one use LLVM  consists of 500+ lines of Cpp code and the other one generating JAVA bytecode consist of 400+ lines of Java code, I think it is safe to say that the implementation complexity of C backend is smaller.
+
+The source code and tests can be found at [Bril2C](https://github.com/xu3kev/bril2c) . The tool to translate Bril JSON format to C is `bril2c.py`. It takes input on stdin and produces output on stdout.
+
+Benchmark
+---
+I create some simple benchmark to measure the timing compared with the interpreter run in Node.js.
+
+Each test was measured for running 1000 times. In order to avoid the startup time of Node.js, I slightly modify the `brili` interpreter to run the Bril program 1000 times internally so I can avoid running `brili` 1000 times.
+
+The experiments are all run on Intel Xeon CPU E5-1630 v3 @ 3.70GHz with Ubuntu 16.04. Turbo boost is turned off and the scaling governor is set to performance.
+
+The version of Node.js to run Bril interpreter is 12.11.0.
+The gcc version  is 5.4.0 and optimization flag is O3.
+
+The benchmark result is as following.
+<br>
+<img src="../../img/c_backend_benchmark.png" width="500">
+<br>
+The four test program are factorial computation, Fibonacci sequence, polynomial multiplication and matrix multiplication.
+We can see that we gain a significant speedup across different tests. However, because Bril now has only bool and int type, the thing we can compute is still limited. In the future, as Bril extends more feature such as floating point arithmetic and array, we can have more practical benchmark.
```

Comment:
```suggestion
We can see that we gain a significant speedup across different tests. However, because Bril now has only bool and int type, the things we can compute are still limited. In the future, as Bril extends with more features such as floating point arithmetic and arrays, we can have more practical benchmarks.
```